### PR TITLE
Revert the parser refactor

### PR DESCRIPTION
**Pull request by danieldk**
## Description

Reverse the parser refactor until we shake out the remaining issues.

### Types of change

Maintenance.

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: spacy/errors.py [hunk 1]
```diff
@@ -215,12 +215,6 @@ class Warnings(metaclass=ErrorsWithCodes):
             "key attribute for vectors, configure it through Vectors(attr=) or "
             "'spacy init vectors --attr'")
 
-    # v4 warning strings
-    W400 = ("`use_upper=False` is ignored, the upper layer is always enabled")
-    W401 = ("`incl_prior is True`, but the selected knowledge base type {kb_type} doesn't support prior probability "
-            "lookups so this setting will be ignored. If your KB does support prior probability lookups, make sure "
-            "to return `True` in `.supports_prior_probs`.")
-
```
[the remark this answers is not in the document]
Comment (danieldk) [reply spacy/errors.py]: Fixed.

File: website/docs/usage/embeddings-transformers.mdx [hunk 2]
```diff
@@ -140,7 +140,7 @@ factory = "tok2vec"
 factory = "ner"
 
 [components.ner.model]
-@architectures = "spacy.TransitionBasedParser.v3"
+@architectures = "spacy.TransitionBasedParser.v1"
```
Comment (adrianeboyd) [inline website/docs/usage/embeddings-transformers.mdx:143]: ? and below

```suggestion
@architectures = "spacy.TransitionBasedParser.v2"
```

Comment (danieldk) [reply website/docs/usage/embeddings-transformers.mdx]: Thanks! Fixed.

File: spacy/compat.py [hunk 3]
```diff
@@ -23,6 +23,11 @@
 except ImportError:
     cupy = None
 
+if sys.version_info[:2] >= (3, 8):  # Python 3.8+
+    from typing import Literal, Protocol, runtime_checkable
+else:
+    from typing_extensions import Literal, Protocol, runtime_checkable  # noqa: F401
+
```
Comment (adrianeboyd) [inline spacy/compat.py:30]: This shouldn't be necessary since we've dropped python 3.7 already.

File: spacy/errors.py [hunk 4]
```diff
@@ -1000,6 +999,7 @@ class Errors(metaclass=ErrorsWithCodes):
     E4011 = ("Server error ({status_code}), couldn't fetch {url}")
 
 
+
```
Comment (adrianeboyd) [inline spacy/errors.py:1002]:
```suggestion
```

File: spacy/ml/models/parser.py [hunk 5]
```diff
@@ -1,66 +1,24 @@
-import warnings
-from typing import Any, List, Literal, Optional, Tuple
+from typing import List, Optional, cast
```
Comment (adrianeboyd) [inline spacy/ml/models/parser.py:1]:
```suggestion
from typing import List, Literal, Optional, cast
```

File: spacy/ml/models/parser.py [hunk 6]
```diff
@@ -1,66 +1,24 @@
...
 
-from thinc.api import Model
+from thinc.api import Linear, Model, chain, list2array, use_ops, zero_init
 from thinc.types import Floats2d
 
-from ...errors import Errors, Warnings
-from ...tokens.doc import Doc
+from ...compat import Literal
```
Comment (adrianeboyd) [inline spacy/ml/models/parser.py:6]:
```suggestion
```

**Review by adrianeboyd — COMMENTED** [review]
We don't need to bring back python <3.8 compat.

File: spacy/compat.py (hunk 3 above)
Comment (danieldk) [reply spacy/compat.py]: Fixed.

File: website/docs/api/cli.mdx [hunk 7]
```diff
@@ -361,7 +361,7 @@ Module     spacy.language
 File       /path/to/spacy/language.py (line 64)
 ℹ [components.ner.model]
 Registry   @architectures
-Name       spacy.TransitionBasedParser.v3
+Name       spacy.TransitionBasedParser.v1
```
Comment (adrianeboyd) [inline website/docs/api/cli.mdx:364]: Also ?

```suggestion
Name       spacy.TransitionBasedParser.v2
```

File: website/docs/api/cli.mdx [hunk 8]
```diff
@@ -371,7 +371,7 @@ Module     spacy.ml.models.tok2vec
 File       /path/to/spacy/ml/models/tok2vec.py (line 16)
 ℹ [components.parser.model]
 Registry   @architectures
-Name       spacy.TransitionBasedParser.v3
+Name       spacy.TransitionBasedParser.v1
```
Comment (adrianeboyd) [inline website/docs/api/cli.mdx:374]:
```suggestion
Name       spacy.TransitionBasedParser.v2
```

File: website/docs/api/cli.mdx (hunk 7 above)
Comment (danieldk) [reply website/docs/api/cli.mdx:364]: I wasn't to sure about that, since it's an example with program output.

**Comment by adrianeboyd** [comment]
This is mildly challenging to review!

I can confirm that the changes to the tests look fine (everything reverted except the distillation tests, right?).

Nothing else jumps out at me at this point, so I think we'll just have to keep this mind when merging in `develop` later. I think in particular we might have to sort out the cython profiling settings a bit, but that should be clear from the python 3.12 tests.

File: website/docs/api/cli.mdx (hunk 7 above)
Comment (adrianeboyd) [reply website/docs/api/cli.mdx]: Ah, you're right. I think this doesn't belong in this PR. (On the other hand, it doesn't matter that much either at this point...)

**Review by svlandeg — APPROVED** [review]
As Adriane states, pretty challenging to review 🙈 

I think we could go ahead and merge though, so we can get through the next steps...

**Pull request merged**